### ISSUE TITLE
pre-validation FileFilters  concept in Checkstyle

### ISSUE THREAD

**Issue by romani** [checkstyle/checkstyle#2866]
splited from #2103 

For me, the second scan produces around 13,000 files. After some examining, CS isn't pre-dropping these files based on my suppressions or even the file extensions specified. Here, https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/Main.java#L378, it only does a plain scan picking everything up in its path. For me, these folders alone are needlessly being deep scanned for 11,000 files that will never be checked.

**Comment by romani** [comment]
existing Suppressions as post-validation AuditEvent filters. We can not examine their configs without braking original purpose of them.

What I see from your request is desire to have file filters to suppress all Checks or may be certain Check before validation process. This is pre-validation filter concept that we do not support currently. Existing filtering by extension that we have in Checker and Treewalker is too simple. 

But definitely need pre-validation filters. I already had a lot of examples when it will be useful. Main reason why it is not yet created as usual lack of time and too much functional problems in our project and moratorium period. I postponed almost all API changes and extensions.

We could start thinking about it now .... .What we need:
1) looks like FileFilter will be interface and should be registered as module
2) We need to predict its usage in plugins (maven, eclispe-cs, ....)
3) We need to predict usage of it in case of global config for all projects in organization/company (this concept "global vs local checkstyle config " is also not ready, but users do complain about it, I thought that could find some time to meditate about this after finishing moratorium period).

**Comment by rnveach** [comment]
> filters to suppress all Checks or may be certain Check before validation process

My original desire was before we even started talking about checks. I am aiming more at the routine for loading possibly files to process (`Main.getFilesToProcess`). Originally, I wanted a way to skip specific directories (bin, .git, .metadata, etc) from being scanned for possible files by checkstyle. These directories are usually pretty big and never contain files usable by checkstyle. Scanning them and filtering out unneeded files was slowing down checkstyle alot for me.

**Comment by romani** [comment]
after introduction of Module system to java, https://github.com/checkstyle/checkstyle/issues/3059 , this issue become acute to let users use java9 even we do not support parsing of such java files.

**Comment by romani** [comment]
issue become approved as it is the only way for us to work on java9 source and  start validation openjdk project again (we need to work on openjdk sources to let openjdk team use us)

**Comment by rnveach** [comment]
@romani I will work on this next.

Here are my thoughts on implementation.
We use a similar format as `suppressions`, but call it `exclusions`. For now it will only support `files` attribute, but we can add more later like `modules` (if a module/check on a specific file happens to throw a NPE between releases).
````
<?xml version="1.0"?>
<!DOCTYPE exclusions PUBLIC
    "-//Puppy Crawl//DTD Exclusions 1.0//EN"
    "http://www.puppycrawl.com/dtds/exclusions_1_0.dtd">

<exclusions>
  <exclude files="path/SomeFile\.ext"/>
</exclusions> 
````
We use pretty much the same code as suppressions to load and process it, but name them with `exclusions`.
Exclusions will have to be added as an external resource for the cache.

I believe all we need to do for now is place the new exclusion logic at `Checker.processFiles` before we call `fireFileStarted` to ignore the file if it matches.
I was originally thinking of making it part of `fireFileStarted` similar to how we have `fireErrors` with suppressions, but this would require an API change in the class and require filters to identify if the event is specifically for them.

What do you think?

**Comment by romani** [comment]
sorry for long delay in reply

> Originally, I wanted a way to skip specific directories (bin, .git, .metadata, etc) from being scanned for possible files by checkstyle 

if you want to fix how `Main.getFilesToProcess` get list of files, we need to have new CLI option, I am completely ok with any, CLI is not our API, so we could do any changes to make CLI usage convenient. We even can load config and only after that do make a list of files (it was done like this only it was easy to implement).
We could split issue to make both if we want.

> We use a similar format as suppressions

using separate files for exclusion is not very good, and only required when amount of exclusions is significant. Do we expect amount to be huge ?
I do not .
User can define list folders and file extensions as singleline , it will not be big for most projects.
So we can have a module with property "excludes" as list of Regexp. 
Extra files for modules it bad design usually. Users do not like additional files. It is better to keep all in one file. There are already bunch of requests to make filter file embeddable to main config.

> I believe all we need to do for now is place the new exclusion logic at Checker.processFiles before we call fireFileStarted to ignore the file if it matches.

yes.

**Comment by rnveach** [comment]
> using separate files for exclusion ... only required when amount of exclusions is significant. 

The 'amount' isn't my main reason for wanting a separate file. It is just for a clear separation between configuration and exclusions, just like with suppressions.
I know people want to be able to combine them into one file, but I like to keep them separate just like we separate different classes in Java.

**Comment by romani** [comment]
fix is merged.
